### ISSUE TITLE
diff: default comparison behaviour

### ISSUE THREAD

**Issue by casperdcl** [iterative/dvc#7578]
from https://github.com/iterative/example-repos-dev/pull/107#pullrequestreview-940619306:

> changing the default behaviour of `dvc metrics diff` to automatically compare with `(REF_unspecified_on_CLI && $GIT_REF == HEAD) ? HEAD~1 : $MAIN_BRANCH`?

(perhaps duplicate issue, only did a quick search)

**Comment by pmrowla** [comment]
This seems like something that belongs in CML to me, I don't think it makes sense for DVC to have code for handling github CI specific env vars in core DVC.

**Comment by casperdcl** [comment]
sorry not referring to env vars, I mean:

```
git checkout master
dvc metrics diff $SOME_REF  # works fine
dvc metics diff # can fail
```

pseudocode logic to fix this on DVC side:

```python
if args.ref is None:
    if repo.current_ref == repo.main_branch:
        args.ref = str(repo.current_ref - 1)
    else:
        args.ref = str(repo.main_branch)
```

**Comment by dberenbaum** [comment]
> ```shell
> git checkout master
> dvc metrics diff $SOME_REF  # works fine
> dvc metics diff # can fail
> ```

@casperdcl Do you know in which scenarios `dvc metrics diff` can fail? The default behavior should compare the current workspace to HEAD. Not sure we want to default to always assuming that we are comparing commits.

**Comment by casperdcl** [comment]
Yes, see https://github.com/iterative/example-repos-dev/pull/105 i.e. "the current workspace" *is* "HEAD"

**Comment by casperdcl** [comment]
ok so we're saying the clunky patch in https://github.com/iterative/example-repos-dev/pull/105 is required, and there's no way DVC could implement a flag (`dvc metrics diff --last` or something) to avoid all the if-then-else bash trickery?

**Comment by daavoo** [comment]
> ok so we're saying the clunky patch in [iterative/example-repos-dev#105](https://github.com/iterative/example-repos-dev/pull/105) is required, and there's no way DVC could implement a flag (`dvc metrics diff --last` or something) to avoid all the if-then-else bash trickery?

I think that in the linked scenario the issue is that the trigger is too generic (`on: push`) while trying to cover 2 different scenarios, semantically different: comparing branch vs main / comparing current commit vs previous.

DVC scope is CLI and both scenarios are covered. It's up to the person setting up the CI context to take care of calling the right arguments. Either with the if/then bash or by having more granular triggers (separate workflows by pushes to the main branch vs. other branches)

**Comment by pmrowla** [comment]
> ok so we're saying the clunky patch in [iterative/example-repos-dev#105](https://github.com/iterative/example-repos-dev/pull/105) is required, and there's no way DVC could implement a flag (`dvc metrics diff --last` or something) to avoid all the if-then-else bash trickery?

I'm also confused as to why the linked PR does different things with the `plots diff` and `metrics diff` calls. Why does `metrics diff` need the extra if/else logic but not `plots diff`? It looks like they should just both be `dvc plots/metrics diff main`?

**Comment by shcheklein** [comment]
> I'm also confused as to why the linked PR does different things with the plots diff and metrics diff calls.

I think, it was fixed here https://github.com/iterative/example-repos-dev/pull/107/files

> Is the point of the example to compare the workspace against a known branch/tag or is it to compare the workspace against the previous git commit, because as @daavoo said, that's two different scenarios.

It does both. If commit belongs to main it would compare it with previous for any PR (branch) it's comparing with main latest.

--------

I doubt we should be making this behavior part of DVC though. It's too specific to this use case, will introduce asymmetry into commands. Sounds too complicated tbh.

**Comment by dberenbaum** [comment]
> If commit belongs to main it would compare it with previous

Is it a common enough scenario to be worth the additional logic? Comparing to the head of main might be enough for most use cases and is more clear than trying to address every possible way that people might want to push and compare changes. For example, a forking workflow probably needs entirely different logic to compare to the upstream fork.

**Comment by shcheklein** [comment]
> It feels beyond a get started scenario to me 🤷 .

Yep, a bit. But it was done to make it look a bit better in Studio (have more than one report) rather than trying to keep very simple for the sake of showing CML itself. I think that's fine in this specific case since it's not a primary entry point into CML.

**Comment by dberenbaum** [comment]
@casperdcl WDYT? Can we close this or move it to a discussion if you still feel there could be something to change in DVC?

**Comment by casperdcl** [comment]
sure, happy to close if we feel it's not something DVC should be simplifying
